src/charm.py

```diff
@@ -977,6 +977,12 @@ def _handle_reconfigure(self, event: UpdateStatusEvent):
         event.unit = self.unit
         self._on_relation_handler(event)
 
+        # make sure all nodes in the replica set have the same priority for re-election. This is
+        # necessary in the case that pre-upgrade hook fails to reset the priority of election for
+        # cluster nodes.
+        with MongoDBConnection(self.mongodb_config) as mongod:
+            mongod.set_replicaset_election_priority(priority=1)
```

Comment:
TL;DR Yes and Yes

Long Story:
The election priority is set + reset when the pre-upgrade-check is ran. It is used to "force" mongodb to move its primary to the lowest unit (Context: MongoDB does not let the user choose which unit the primary is on). If for some reason the charm fails to reset the election priority during the pre-upgrade-check hook and the user decides to move forward with `juju refresh` (despite the failing event) the charm will be configured to always have elections result in the lowest unit as a primary which does not follow HA-practices. It is necessary that we try to recorrect this and "self heal" this cluster config

Reply:
> okay so, to confirm, if there's a primary switchover during an upgrade to a unit that's higher than the lowest unit number—that's acceptable (but you want to avoid it)?

Not quite, we ant to avoid that the `set_replicaset_election_priority ` never gets rest

If a rollback occurs, the pre-upgrade hook will run and reset the primary.

Reply:
Hmmm, definitely an interesting case you outline. The case being, before upgrading we move primary to the last unit to upgrade, but then the upgrade can last weeks and the primary moves to a different unit. TBH I am not sure here. 🤔 

On one hand: I believe that if a primary election happened, then there was a good reason for it. i.e. secondaries lost connection with the Primary. I don't think it makes sense for us to force a unit to be primary if internal mongodb "feels" otherwise. 

On the other hand: we want the primary to be the last upgrade, so we should try to guarantee that if possible. 

I am not sure if we can tell why MongoDB decided to elect a different replica as primary, which would be useful. I don't think we should "force" the charm to have a certain primary if the internal daemons think otherwise. The entire purpose of primary elections is to maintain HA, and if we override it we risk decreasing performance of the cluster. **But then again, we really want the primary to be the last unit to upgrade.** and maybe after enough time passes the cluster/replica set is "settled" and it can infact be moved back.

Personally, I think this is a complicated case to handle and I think it would be best to leave the code is. i.e. if the internal mongod daemon decides to move the primary to a different unit partway through the upgrade we should let it, since it probably has a good reason. 

But I would like to get someone else's opinion who might have more MongoDB knowledge (i.e. @delgod)

Reply:
This is pending a technical conversation - we will revisit this topic later, if we decide to implement logic like this we will do so in a separate PR as it is regarding the upgrade procedure and not pre-upgrade-check

Thank you for the thought provoking topic!

Reply:
> want or need? wondering if it's a hard requirement or if the workload itself will be fine if primary is not last to upgrade

I believe the next
e.g. upgrade from ver.X to ver.Y
1) We prefer, if possible, to have a minimal amount of switchovers during the upgrade.
2) It is acceptable if, because of failover, the primary ver.X is switched over to another ver.X member.
3) it is acceptable if, because of failover, the primary ver.X is switched over to ver.Y member **only when no ver.X members are left alive/available**.

Reply:
According [to the mongodb docs](https://www.mongodb.com/docs/manual/core/replica-set-elections/), an election will occur for the following reasons:
- Adding a new node to the replica set [We block this from happening during an upgrade]
- [initiating a replica set,](https://www.mongodb.com/docs/manual/reference/method/rs.initiate/#mongodb-method-rs.initiate) [This will not occur during an upgrade]
- performing replica set maintenance using methods such as [rs.stepDown()](https://www.mongodb.com/docs/manual/reference/method/rs.stepDown/#mongodb-method-rs.stepDown) or [rs.reconfig()](https://www.mongodb.com/docs/manual/reference/method/rs.reconfig/#mongodb-method-rs.reconfig) - [rs.stepDown() will not occur during an upgrade unless we willingly move primary + rs.reconfig() occurs only in the case of self-healing]
- the [secondary](https://www.mongodb.com/docs/manual/reference/glossary/#std-term-secondary) members losing connectivity to the primary for more than the configured [timeout](https://www.mongodb.com/docs/manual/reference/replica-configuration/#mongodb-rsconf-rsconf.settings.electionTimeoutMillis) (10 seconds by default). [this is a part of self healing and will occur during upgrades, but is expected]

So I believe it is fine to keep it as is.